src/main/scala/org/constellation/util/wallet/CreateNewTransaction.scala

```diff
@@ -0,0 +1,107 @@
+package org.constellation.util.wallet
+
+import java.io._
+import java.security._
+
+import cats.data.EitherT
+import cats.effect.{ExitCode, IO, IOApp, Sync}
+import constellation._
+import org.constellation.domain.transaction.{LastTransactionRef, TransactionService}
+import org.constellation.keytool.{KeyStoreUtils, KeyUtils}
+import org.constellation.primitives.Transaction
+import scopt.OParser
+
+import scala.util.Try
+
+object CreateNewTransaction extends IOApp {
+  /*
+  Note: these vals need type annotation to compile
+   */
+  val transactionParser: FileInputStream => IO[Option[Transaction]] =
+    KeyStoreUtils.parseFileOfTypeOp[IO, Transaction](ParseExt(_).x[Transaction])
+
+  val transactionWriter: Transaction => FileOutputStream => IO[Unit] =
+    KeyStoreUtils.storeTypeWithFileStream[IO, Transaction](SerExt(_).json)
+
+  def run(args: List[String]): IO[ExitCode] = {
+    for {
+      cliParams <- loadCliParams[IO](args)
+      kp <- loadKeyPairFrom[IO](cliParams)
+      prevTransactionOp <- KeyStoreUtils.readFromFileStream[IO, Option[Transaction]](cliParams.accountPath,
+                                                                                     transactionParser)
+      transactionEdge = TransactionService.createTransactionEdge(
+        KeyUtils.publicKeyToAddressString(kp.getPublic),
+        cliParams.destination,
+        cliParams.amount.toDouble.toLong,
+        kp
+      )
+      transaction = Transaction(transactionEdge, prevTransactionOp.map(_.lastTxRef).getOrElse(LastTransactionRef.empty))
+      transactionWriteBuffer = transactionWriter(transaction)
+      _ <- KeyStoreUtils.storeWithFileStream[IO](cliParams.storePath, transactionWriteBuffer)
+    } yield transaction
+  }.fold[ExitCode](throw _, _ => ExitCode.Success)
+
+  //todo add case for storepass keypass via env
+  def loadKeyPairFrom[F[_]: Sync](cliParams: WalletCliConfig): EitherT[F, Throwable, KeyPair] =
+    if (cliParams.privateKeyStr == null)
+      KeyStoreUtils
+        .keyPairFromStorePath[F](cliParams.keystore, cliParams.alias, cliParams.storepass, cliParams.keypass)
+    else {
+      val kp = KeyUtils.keyPairFromPemStr(cliParams.privateKeyStr, cliParams.pubKeyStr)
+      val eitherLoadOrThrow =
+        Try(Right(kp)).getOrElse(Left(new Throwable("Couldn't load KeyPair with PrivateKey provided")))
+
+      EitherT(Sync[F].delay { eitherLoadOrThrow })
```

Comment:
Classic `null` checks should be avoided and `Option` should be used instead. You can just fold it and lift to EitherT context:

```scala
Option(cliParams.privateKeyString)
   .fold(
      KeyStoreUtils
        .keyPairFromStorePath[F](cliParams.keystore, cliParams.alias, cliParams.storepass, cliParams.keypass),
      Try(KeyUtils.keyPairFromPemStr(cliParams.privateKeyStr, cliParams.pubKeyStr))
          .toEither
          .toEitherT
          .leftMap(_ => new Throwable("Couldn't load KeyPair with PrivateKey provided")),
   )
```